### PR TITLE
Add support build image

### PR DESCRIPTION
**Pull request by jimmyxian**
https://github.com/docker/swarm/issues/632 swarm need this Interface

test code
```
package main

import (
        "archive/tar"
        "bytes"
        "fmt"
        "github.com/samalba/dockerclient"
        "io/ioutil"
)

func main() {
        // Init the client
        docker, _ := dockerclient.NewDockerClient("unix:///var/run/docker.sock", nil)
        files := make(map[string]string)
        files["Dockerfile"] = "FROM busybox"
        buf := new(bytes.Buffer)
        tw := tar.NewWriter(buf)
        for k, v := range files {
                hdr := &tar.Header{
                        Name: k,
                        Size: int64(len(v)),
                }
                err := tw.WriteHeader(hdr)
                if err != nil {
                        return
                }
                _, err = tw.Write([]byte(v))
                if err != nil {
                        return
                }
        }
        image := dockerclient.BuildImage{
                Config:         nil,
                DockerfileName: "",
                Context:        buf,
                RemoteURL:      "",
                RepoName:       "testimage:ok",
                SuppressOutput: false,
                NoCache:        false,
                Remove:         false,
                ForceRemove:    false,
                Pull:           true,
                Memory:         0,
                MemorySwap:     0,
                CpuShares:      0,
                CpuPeriod:      0,
                CpuQuota:       0,
                CpuSetCpus:     "",
                CpuSetMems:     "",
                CgroupParent:   "",
        }
        reader, err := docker.BuildImage(image)
        if err != nil {
                fmt.Printf("%s", err.Error())
        }
        data, _ := ioutil.ReadAll(reader)
        fmt.Printf("%s", data)
}


```

Signed-off-by: Xian Chaobo <xianchaobo@huawei.com>

### CODE REVIEW

**Comment by jimmyxian** [comment]
ping @donhcd @vieux @aluzzardi @ehazlett

File: dockerclient.go [hunk 1]
```diff
@@ -505,3 +505,53 @@ func (client *DockerClient) ImportImage(source string, repository string, tag st
...
+		v.Set("forcerm", "1")
+	}
+	if image.Quiet {
+		v.Set("q", "1")
+	}
+
+	v.Set("cpusetcpus", image.CpuSet)
+	v.Set("cpushares", strconv.FormatInt(image.CpuShares, 10))
+	v.Set("memory", strconv.FormatInt(image.Memory, 10))
+	v.Set("memswap", strconv.FormatInt(image.MemorySwap, 10))
```
Comment (donhcd) [inline dockerclient.go]: nit: use `strconv.Atoi` for these 3?

File: dockerclient.go [hunk 2]
```diff
@@ -505,3 +505,53 @@ func (client *DockerClient) ImportImage(source string, repository string, tag st
...
+
+	headers := make(map[string]string)
+	if image.Auth != nil {
+		headers["X-Registry-Auth"] = image.Auth.encode()
```
Comment (donhcd) [inline dockerclient.go]: why is this here? I haven't looked at the docker source code for this but the docs say that you should provide `X-Registry-Config`, and the `ConfigFile` already contains `AuthConfig`s

File: auth.go [hunk 3]
```diff
@@ -19,3 +19,16 @@ func (c *AuthConfig) encode() string {
...
+// ConfigFile holds parameters for authenticating during a BuildImage request
+type ConfigFile struct {
+	Configs  map[string]AuthConfig `json:"configs,omitempty"`
+	rootPath string
+}
+
+// encode the configuration struct into base64 for the X-Registry-Config header
+func (c *ConfigFile) encode() string {
+	var buf bytes.Buffer
+	json.NewEncoder(&buf).Encode(c)
```
Comment (donhcd) [inline auth.go]: I think it's best to check for the error here, and make this method return `(string, error)`

File: dockerclient.go (hunk 1 above)
Comment (jimmyxian) [inline dockerclient.go]: OK

File: dockerclient.go (hunk 2 above)
Comment (jimmyxian) [inline dockerclient.go]: https://github.com/docker/docker/blob/v1.6.1/api/server/server.go#L1024
Yeah, in docker master, it has already removed. But in 1.6.1 or before 1.6.1 it has the `AuthConfig`. 
Should we removed in dockerclient?

File: auth.go (hunk 3 above)
Comment (jimmyxian) [inline auth.go]: OK

File: dockerclient.go (hunk 1 above)
Comment (jimmyxian) [inline dockerclient.go]: Here, want a string, so use `FormatInt`  Int to String
`Atoi` String to Int

Comment (donhcd) [inline dockerclient.go]: Sorry, meant Itoa

**Comment by donhcd** [comment]
github hid this but it's still relevant:
screen shot 2015-05-18 at 16 19 58

thoughts @aluzzardi @vieux @samalba ?

**Comment by ehazlett** [comment]
Any update on this?  I think it would be cool to have build capabilities.

**Comment by jimmyxian** [comment]
@ehazlett 
Yeah, swarm 0.4 also need this.
Anyway, I have compared the latest docker code, find out that more fields is added in build. so I will update this recently.
Also, any thought about `AuthConfig` that @donhcd says?

**Comment by donhcd** [comment]
imo we can remove the `X-Registry-Auth` header and the `Auth` field in `BuildImage`

looks like `X-Registry-Config` has been used since 1.0 https://github.com/docker/docker/blob/v1.0.0/api/server/server.go#L888-L891

**Comment by jimmyxian** [comment]
ping @donhcd @ehazlett 
Updated and check again. Thanks.

File: dockerclient.go [hunk 4]
```diff
@@ -505,3 +509,61 @@ func (client *DockerClient) ImportImage(source string, repository string, tag st
...
+	if image.RemoteURL != "" {
+		v.Set("remote", image.RemoteURL)
+	}
+	if image.NoCache {
+		v.Set("nocache", "1")
+	}
+	if image.Pull {
+		v.Set("pull", "1")
+	}
+	if image.Remove {
```
Comment (donhcd) [inline dockerclient.go]: this is repeated

Comment (jimmyxian) [inline dockerclient.go]: Yeah, I will fix this now

Comment (jimmyxian) [inline dockerclient.go]: Done! review again.thanks

**Comment by jimmyxian** [comment]
ping @vieux @aluzzardi  @donhcd
Need code review again, Thanks

**Comment by donhcd** [comment]
LGTM

**Comment by donhcd** [comment]
ping @ehazlett

**Comment by ehazlett** [comment]
LGTM
